File: output/java/tests/39_tagged_instantiation.java
import java.util.*;


import java.util.*;

class CalculatorFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    CalculatorFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    CalculatorFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class CalculatorFrameContext {
    CalculatorFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    CalculatorFrameContext(CalculatorFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class CalculatorCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    CalculatorFrameEvent forward_event;
    CalculatorCompartment parent_compartment;

    CalculatorCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    CalculatorCompartment copy() {
        CalculatorCompartment c = new CalculatorCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class Calculator {
    private ArrayList<CalculatorCompartment> _state_stack;
    private CalculatorCompartment __compartment;
    private CalculatorCompartment __next_compartment;
    private ArrayList<CalculatorFrameContext> _context_stack;
    public int result = 0;

    public Calculator() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new CalculatorCompartment("Ready");
        __next_compartment = null;
        CalculatorFrameEvent __frame_event = new CalculatorFrameEvent("$>");
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(CalculatorFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            CalculatorFrameEvent exit_event = new CalculatorFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                CalculatorFrameEvent enter_event = new CalculatorFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    CalculatorFrameEvent enter_event = new CalculatorFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(CalculatorFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("Ready")) {
            _state_Ready(__e);
        }
    }

    private void __transition(CalculatorCompartment next) {
        __next_compartment = next;
    }

    public int add(int a, int b) {
        HashMap<String, Object> __params = new HashMap<>();
        __params.put("a", a);
        __params.put("b", b);
        CalculatorFrameEvent __e = new CalculatorFrameEvent("add", __params);
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public int get_result() {
        CalculatorFrameEvent __e = new CalculatorFrameEvent("get_result");
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    private void _state_Ready(CalculatorFrameEvent __e) {
        if (__e._message.equals("add")) {
            var a = (int) __e._parameters.get("a");
            var b = (int) __e._parameters.get("b");
            this.result = a + b;
            _context_stack.get(_context_stack.size() - 1)._return = this.result;
            return;
        } else if (__e._message.equals("get_result")) {
            _context_stack.get(_context_stack.size() - 1)._return = this.result;
            return;
        }
    }
}

class CounterFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    CounterFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    CounterFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class CounterFrameContext {
    CounterFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    CounterFrameContext(CounterFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class CounterCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    CounterFrameEvent forward_event;
    CounterCompartment parent_compartment;

    CounterCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    CounterCompartment copy() {
        CounterCompartment c = new CounterCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class Counter {
    private ArrayList<CounterCompartment> _state_stack;
    private CounterCompartment __compartment;
    private CounterCompartment __next_compartment;
    private ArrayList<CounterFrameContext> _context_stack;
    public int count = 0;

    public Counter() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new CounterCompartment("Active");
        __next_compartment = null;
        CounterFrameEvent __frame_event = new CounterFrameEvent("$>");
        CounterFrameContext __ctx = new CounterFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(CounterFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            CounterFrameEvent exit_event = new CounterFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                CounterFrameEvent enter_event = new CounterFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    CounterFrameEvent enter_event = new CounterFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(CounterFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("Active")) {
            _state_Active(__e);
        }
    }

    private void __transition(CounterCompartment next) {
        __next_compartment = next;
    }

    public void increment() {
        CounterFrameEvent __e = new CounterFrameEvent("increment");
        CounterFrameContext __ctx = new CounterFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (void) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public int get_count() {
        CounterFrameEvent __e = new CounterFrameEvent("get_count");
        CounterFrameContext __ctx = new CounterFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    private void _state_Active(CounterFrameEvent __e) {
        if (__e._message.equals("get_count")) {
            _context_stack.get(_context_stack.size() - 1)._return = this.count;
            return;
        } else if (__e._message.equals("increment")) {
            this.count = this.count + 1;
        }
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("=== Test 39: Tagged System Instantiation ===");

        // Tagged instantiation - validated at transpile time
        var calc = new Calculator();
        var counter = new Counter();

        // Test Calculator
        int result = calc.add(3, 4);
        if (result != 7) {
            System.out.println("FAIL: Expected 7, got " + result);
            System.exit(1);
        }
        System.out.println("Calculator.add(3, 4) = " + result);

        result = calc.get_result();
        if (result != 7) {
            System.out.println("FAIL: Expected 7, got " + result);
            System.exit(1);
        }
        System.out.println("Calculator.get_result() = " + result);

        // Test Counter
        counter.increment();
        counter.increment();
        counter.increment();
        int count = counter.get_count();
        if (count != 3) {
            System.out.println("FAIL: Expected 3, got " + count);
            System.exit(1);
        }
        System.out.println("Counter after 3 increments: " + count);

        System.out.println("PASS: Tagged instantiation works correctly");
    }
}

File: output/go/tests/39_tagged_instantiation.go

package main

import "fmt"
import "os"

type CalculatorFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type CalculatorFrameContext struct {
    _event  CalculatorFrameEvent
    _return any
    _data   map[string]any
}

type CalculatorCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *CalculatorFrameEvent
    parentCompartment *CalculatorCompartment
}

func newCalculatorCompartment(state string) *CalculatorCompartment {
    return &CalculatorCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *CalculatorCompartment) copy() *CalculatorCompartment {
    nc := &CalculatorCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type Calculator struct {
    _state_stack []*CalculatorCompartment
    __compartment *CalculatorCompartment
    __next_compartment *CalculatorCompartment
    _context_stack []CalculatorFrameContext
    result int
}

func NewCalculator() *Calculator {
    s := &Calculator{}
    s._state_stack = make([]*CalculatorCompartment, 0)
    s._context_stack = make([]CalculatorFrameContext, 0)
    s.__compartment = newCalculatorCompartment("Ready")
    s.__next_compartment = nil
    __frame_event := CalculatorFrameEvent{_message: "$>", _parameters: nil}
    __ctx := CalculatorFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *Calculator) __kernel(__e *CalculatorFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &CalculatorFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &CalculatorFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &CalculatorFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *Calculator) __router(__e *CalculatorFrameEvent) {
    switch s.__compartment.state {
    case "Ready":
        s._state_Ready(__e)
    }
}

func (s *Calculator) __transition(next *CalculatorCompartment) {
    s.__next_compartment = next
}

func (s *Calculator) Add(a int, b int) int {
    __params := map[string]any{
        "a": a,
        "b": b,
    }
    __e := CalculatorFrameEvent{_message: "Add", _parameters: __params}
    __ctx := CalculatorFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *Calculator) GetResult() int {
    __e := CalculatorFrameEvent{_message: "GetResult"}
    __ctx := CalculatorFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *Calculator) _state_Ready(__e *CalculatorFrameEvent) {
    if __e._message == "Add" {
        a := __e._parameters["a"].(int)
        _ = a
        b := __e._parameters["b"].(int)
        _ = b
        s.result = a + b
        s._context_stack[len(s._context_stack)-1]._return = s.result
        return
    } else if __e._message == "GetResult" {
        s._context_stack[len(s._context_stack)-1]._return = s.result
        return
    }
}

type CounterFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type CounterFrameContext struct {
    _event  CounterFrameEvent
    _return any
    _data   map[string]any
}

type CounterCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *CounterFrameEvent
    parentCompartment *CounterCompartment
}

func newCounterCompartment(state string) *CounterCompartment {
    return &CounterCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *CounterCompartment) copy() *CounterCompartment {
    nc := &CounterCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type Counter struct {
    _state_stack []*CounterCompartment
    __compartment *CounterCompartment
    __next_compartment *CounterCompartment
    _context_stack []CounterFrameContext
    count int
}

func NewCounter() *Counter {
    s := &Counter{}
    s._state_stack = make([]*CounterCompartment, 0)
    s._context_stack = make([]CounterFrameContext, 0)
    s.__compartment = newCounterCompartment("Active")
    s.__next_compartment = nil
    __frame_event := CounterFrameEvent{_message: "$>", _parameters: nil}
    __ctx := CounterFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *Counter) __kernel(__e *CounterFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &CounterFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &CounterFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &CounterFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *Counter) __router(__e *CounterFrameEvent) {
    switch s.__compartment.state {
    case "Active":
        s._state_Active(__e)
    }
}

func (s *Counter) __transition(next *CounterCompartment) {
    s.__next_compartment = next
}

func (s *Counter) Increment() {
    __e := CounterFrameEvent{_message: "Increment"}
    __ctx := CounterFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *Counter) GetCount() int {
    __e := CounterFrameEvent{_message: "GetCount"}
    __ctx := CounterFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *Counter) _state_Active(__e *CounterFrameEvent) {
    if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = s.count
        return
    } else if __e._message == "Increment" {
        s.count = s.count + 1
    }
}

func main() {
	fmt.Println("=== Test 39: Tagged System Instantiation ===")

	// Tagged instantiation - validated at transpile time
	calc := NewCalculator()
	counter := NewCounter()

	// Test Calculator
	result := calc.Add(3, 4)
	if result != 7 {
		fmt.Printf("FAIL: Expected 7, got %d\n", result)
		os.Exit(1)
	}
	fmt.Printf("Calculator.Add(3, 4) = %d\n", result)

	result = calc.GetResult()
	if result != 7 {
		fmt.Printf("FAIL: Expected 7, got %d\n", result)
		os.Exit(1)
	}
	fmt.Printf("Calculator.GetResult() = %d\n", result)

	// Test Counter
	counter.Increment()
	counter.Increment()
	counter.Increment()
	count := counter.GetCount()
	if count != 3 {
		fmt.Printf("FAIL: Expected 3, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Counter after 3 increments: %d\n", count)

	fmt.Println("PASS: Tagged instantiation works correctly")
}
